README.md
![Pry Theme][logo]

![Pry Theme version badge][ver-badge] [![Build Status][ci-badge]][ci-link]

* Repository: [https://github.com/kyrylo/pry-theme/][pt]
* Wiki: [https://github.com/kyrylo/pry-theme/wiki][wiki]

Description
-----------

Pry Theme is a plugin for [Pry][pry], which helps you to customise your Pry
colors via `prytheme.rb` files.

![Railscasts](./screenshots/railscasts.png)
![Solarized](./screenshots/solarized.png)
![Tomorrow](./screenshots/tomorrow.png)
![Zenburn](./screenshots/zenburn.png)

Installation
------------

All you need is to install the gem. The `pry-theme` plugin will be detected and
used automatically.

    gem install pry-theme

Synopsis
--------

### Theme files

Theme file is nothing but a Ruby file, which has `.prytheme.rb` extension (for
example, `beautiful.prytheme.rb`). In order to set up the desired theme, add the
following line to your `.pryrc`:

    Pry.config.theme = "theme-name"

The default theme is the one from the `pry-classic` family. It is dependent on
your terminal capabilities. For example, if you're using Windows, your default
theme will be `pry-classic-16`, since the Windows terminal can't support more
than 16 colours. If you're using xterm or urxvt, then your default theme will be
`pry-classic-256` (basically, you shouldn't notice it, because it resembles the
default outlook of Pry, as though Pry Theme isn't present). This outlines the
rule: the more colours your terminal supports, the more vivid version of
`pry-classic` is used.

Let's change it to something more neoteric:

    Pry.config.theme = "pry-modern-256"

That's all! Launch your Pry and you will see the changes.

### CLI

Pry Theme has a command-line interface available via Pry. Just launch Pry and
start working with it. For example, you can _temporary_ change themes on the
fly (only for the current session):

    [1] pry(main)> pry-theme try pry-classic-8

This subcommand would switch your current theme to `pry-classic-8` theme.

You can find [more information about CLI in Pry Theme Wiki][cli].

### Managing themes

Creating new themes isn't hard. [Check out Pry Theme Wiki article on that][new_theme].

Theme files must have `.prytheme.rb` extension. Check out [Pry Theme Collection][ptc]
if you want to find some themes other than default ones.

If you already have your theme stored somewhere on disk, just put it in the
`$HOME/.pry/themes` directory.

If you don't want to bother with routine operations, you can install a theme
from the Collection with help of `pry-theme install <name>` subcommand. For
example, you can want to install the xoria256 theme. Just execute
`pry-theme install xoria256` and you're done.

Oh, and don't forget to adjust your `.pryrc`!

Limitations
-----------

No limitations. Pry Theme will run everywhere where you can launch Pry.

Credits
-------

* Thanks to [banister][johndogg] for bringing the plugin in masses and
  contributing a bunch of themes;
* Thanks to Karandashev for "Puzzle" font;
* Thanks to Creatica for "Dited" font;
* Thanks to [noprompt][noprompt] for a HEX to ANSI conversion Ruby
  implementation, which I borrowed from one of his projects.

License
-------

The project uses Zlib License. See LICENSE file for more information.

[pt]: https://github.com/kyrylo/pry-theme/ "Home page"
[ver-badge]: https://badge.fury.io/rb/pry-theme.png "Pry Theme version badge"
[ci-badge]: https://travis-ci.org/kyrylo/pry-theme.png?branch=master "Build status"
[ci-link]: https://travis-ci.org/kyrylo/pry-theme/ "Build history"
[logo]: http://img-fotki.yandex.ru/get/5107/98991937.a/0_7c6c8_871a1842_orig "Pry Theme"
[pry]: https://github.com/pry/pry/ "Pry's home page"
[new_theme]: https://github.com/kyrylo/pry-theme/wiki/Creating-a-New-Theme
[cli]: https://github.com/kyrylo/pry-theme/wiki/Pry-Theme-CLI
[wiki]: https://github.com/kyrylo/pry-theme/wiki
[ptc]: https://github.com/kyrylo/pry-theme-collection
[johndogg]: https://github.com/banister/ "John Dogg"
[noprompt]: https://github.com/noprompt/ "Joel Holdbrooks"


CHANGELOG.md
Pry Theme changelog
===================

### master

### v1.3.1 (August 19, 2020)

* Fixed ``NoMethodError: undefined method `windows?``
  ([#62](https://github.com/kyrylo/pry-theme/pull/62))

### v1.3.0 (January 26, 2019)

* Fixed the "method BaseHelpers#windows? is deprecated" warning
  ([#55](https://github.com/kyrylo/pry-theme/pull/55))

### v1.2.0 (January 16, 2016)

* **IMPORTANT:** dropped support for Ruby <2.0
  ([#51](https://github.com/kyrylo/pry-theme/pull/51))
* Started depending on stock JSON library (no more dependency on the `json` gem)
  ([#49](https://github.com/kyrylo/pry-theme/pull/51))
* Fixed warnings on Ruby 2.4.0+ with regard to Fixnum
  ([#50](https://github.com/kyrylo/pry-theme/pull/51))

### v1.1.3 (July 21, 2014)

* Re-fixed broken behaviour of Pry. It's a temporary hack, to be removed on the
  new Pry release [→](https://github.com/kyrylo/pry-theme/issues/40)
  [→](https://github.com/kyrylo/pry-theme/issues/42)

### v1.1.2 (June 16, 2014)

* Fixed broken behaviour when the Pry Rescue plugin isn't installed along with
  Pry Theme [→](https://github.com/kyrylo/pry-theme/issues/40)

### v1.1.1 (June 14, 2014)

* Fixed error message when `theme_options` is not set at
  all. [→](https://github.com/kyrylo/pry-theme/issues/41)

### v1.1.0 (June 13, 2014)

* Added `theme_options` config hash. It is a general config for Pry Theme
* Added option that enables painting hash keys as symbols.
  [→](https://github.com/kyrylo/pry-theme/issues/30)

  ```ruby
  {foo: 1, :bar => 2}
  #  ^       ^
  #  |       |
  # key    symbol
  ```

  By default `foo` and `bar` have different colours. If you put this inside your
  `.pryrc`, then the key will have the same colour as the symbol.

  ```ruby
  Pry.config.theme_options = {:paint_key_as_symbol => true}
  ```

### v1.0.3 (May 12, 2014)

* Fixed error when Pry.config.theme is set to a theme that doesn't exist
  [→](https://github.com/kyrylo/pry-theme/issues/39)
* Corrected README.md errors and confusing parts

### v1.0.2 (January 28, 2014)

* Set stricter dependencies (namely, Pry Theme started demanding CodeRay v1.1.0
  or higher)

### v1.0.1 (November 18, 2013)

* Fixed `pry-theme convert` and `pry-theme colors`

### v1.0.0 (November 13, 2013)

* **IMPORTANT**: This version is incompatible with CodeRay 1.0.9 and lower
* Fixed CodeRay 1.1.0 support

### v0.2.0 (March 26, 2013)

* **HOT**: Added Windows support
* **NOT SO HOT**: The plugin was rewritten from scratch
* The plugin works everywhere where you can launch Pry
* Improved `vim-detailed` theme
* Added support for 16 colour themes
* Introduced the new theme syntax
* Lots of rewordings, improvements of help messages
* Added fallback mode. It is useful when you are running Pry on poor terminals
  and trying to use a 256 colour theme. It is quite numpty, though.
* Converted switches to subcommands
* Added basic tab completion for installed themes
* Added support for italic font. Does not work eveywhere. For example, I use
  urxvt and it does support italic fonts.
* **NOT BAD**: Added a bunch of new themes

    * pry-monochrome
    * pry-modern-256 (old `pry-modern`)
    * pry-classic-256 (old `pry-classic`; default theme on 256 colour capable
      terminals)

    Including Windows friendly themes:

    * pry-classic-16 (default theme on Windows)
    * pry-classic-8
    * pry-zealand-16 (greenish)
    * pry-zealand-8
    * pry-love-16 (redish)
    * pry-love-8
    * pry-sibera-16 (bluish)
    * pry-sibera-8
    * pry-tepid-16 (yellowish)
    * pry-tepid-8
    * pry-modern-16
    * pry-modern-8
* And probably something else. I could forget something!

And do not forget to check out [the new CLI][cli]!

### v0.1.3 (September 21, 2012)

* Improved integration with [pry-rescue][prescue] plugin.

### v0.1.2 (August 31, 2012)

* Removed uninstaller. Actually, adding it, was a terrible idea. Remove theme
  files by hand, from now on.

### v0.1.1 (August 1, 2012)

* Fixed bug in `--test` command not displaying current theme name;
* Added new Pry Theme attributes: `escape`, `inline_delimiter` and `char`;
* Amended output of `--test` command in order to reflect new attributes.

### v0.1.0 (July 23, 2012)

* **HOT**: Boosted the load speed of the plugin:

  Before:
  ![Before](http://img-fotki.yandex.ru/get/6405/98991937.b/0_7f768_24d92170_orig)

        % time pry -e 'exit'
        0.75s user 0.06s system 94% cpu 0.847 total

  After:
  ![After](http://img-fotki.yandex.ru/get/6400/98991937.b/0_7f767_7c1ad4e9_orig)

        % time pry -e 'exit'
        0.69s user 0.04s system 94% cpu 0.773 total

* **WARM**: added `--edit` command (`-e`). `-e` brings new experience in
  creating themes. With this command you can create and edit themes, both. For
  example, you can create a new theme like this: `pry-theme -e +my-theme`. Check
  out [a wiki entry][cli] for more information;

* **TEPID**: improved `--color` command (`-c`). In this version `-c` can
  understand HEX, RGB, ANSI and human-readable colors. Interested in how to use
  it? Check out [Pry Theme CLI][cli] article in wiki;

* Added [Pry Theme Cheatsheet][cheatsheet] to Pry Theme Wiki;
* Fixed output for `--all-colors`, `--test`, `--list` and `--remote--list`. It
  was hard-coded to `$stdout`, so therefore, when it wasn't `$stdout`,
  information would be printed in the wrong place;
* Fixed duplicate colors 86:aquamarine03 and 122:aquamarine03;
* Fixed a bug when a user doesn't specify `Pry.config.theme` and tries to run
  `pry-theme -l`, which results in a TypeError;
* Added two new 8-color default themes for a users with limited terminals:
  "vim-default" and "vim-detailed";
* Changed output for `--test`: it's more terse and concise now;
* Changed output for `--list`, because the older one looked really silly;
* Removed some spurious attributes for `.prytheme` files;
* Fixed the name of theme "saturday".

### v0.0.10 (July 07, 2012)

* Themes can use any color from 256-color palette as background color. Before

lib/pry-theme/color.rb
module PryTheme
  class Color

    # Stores all possible colours for foregrounds and backgrounds, for 256, 16
    # and 8 colour palettes.
    @@colors = {
      :foreground => {
        256 => {
          'black'                => [0, 16],
          'maroon'               => 1,
          'toad_in_love'         => 2,
          'olive'                => 3,
          'navy_blue'            => 4,
          'violet_eggplant'      => 5,
          'teal'                 => 6,
          'silver'               => 7,
          'gray'                 => [8, 244],
          'red'                  => [9, 196],
          'green'                => [10, 46],
          'yellow'               => [11, 226],
          'blue'                 => [12, 21],
          'magenta'              => [13, 201],
          'cyan'                 => [14, 51],
          'white'                => [15, 231],
          'dark_indigo'          => 17,
          'ultramarine01'        => 18,
          'ultramarine02'        => 19,
          'persian_blue01'       => 20,
          'dark_spring_green'    => 22,
          'dark_turquoise'       => 23,
          'cerulean_grey01'      => 24,
          'denim01'              => 25,
          'royal_blue01'         => 26,
          'royal_blue02'         => 27,
          'toad_in_love01'       => 28,
          'sea_green01'          => 29,
          'teal01'               => 30,
          'cerulean_grey02'      => 31,
          'klein_blue'           => 32,
          'azure01'              => 33,
          'vert_de_pomme01'      => 34,
          'jade01'               => 35,
          'jade02'               => 36,
          'robin_egg_blue01'     => 37,
          'bondi_blue'           => 38,
          'light_blue01'         => 39,
          'vert_de_pomme02'      => 40,
          'malachite01'          => 41,
          'emerald01'            => 42,
          'turquoise'            => 43,
          'robin_egg_blue02'     => 44,
          'electric01'           => 45,
          'spring_green01'       => 47,
          'spring_green02'       => 48,
          'mint_green01'         => 49,
          'aquamarine01'         => 50,
          'flea_belly'           => 52,
          'plum'                 => 53,
          'indigo'               => 54,
          'purple01'             => 55,
          'violet01'             => 56,
          'persian_blue02'       => 57,
          'khaki01'              => 58,
          'wet_asphalt01'        => 59,
          'seroburomalinovyj01'  => 60,
          'denim02'              => 61,
          'royal_blue03'         => 62,
          'royal_blue04'         => 63,
          'olive_drab'           => 64,
          'fern_green'           => 65,
          'slate_gray'           => 66,
          'steel_blue'           => 67,
          'cornflower_blue01'    => 68,
          'azure02'              => 69,
          'grass01'              => 70,
          'emerald02'            => 71,
          'sea_green02'          => 72,
          'robin_egg_blue03'     => 73,
          'bluish01'             => 74,
          'light_blue02'         => 75,
          'vert_de_pomme03'      => 76,
          'pale_green01'         => 77,
          'emerald03'            => 78,
          'aquamarine02'         => 79,
          'robin_egg_blue04'     => 80,
          'sky01'                => 81,
          'bright_green'         => 82,
          'malachite02'          => 83,
          'spring_green03'       => 84,
          'chartreuse01'         => 85,
          'aquamarine03'         => 86,
          'cyan01'               => 87,
          'maroon01'             => 88,
          'eggplant01'           => 89,
          'violet_eggplant01'    => 90,
          'purple02'             => 91,
          'violet02'             => 92,
          'violet03'             => 93,
          'khaki02'              => 94,
          'pale_mauve01'         => 95,
          'seroburomalinovyj02'  => 96,
          'amethyst01'           => 97,
          'amethyst02'           => 98,
          'heliotrope01'         => 99,
          'olive01'              => 100,
          'dark_tan'             => 101,
          'gray01'               => [102, 243],
          'bluish02'             => 103,
          'cornflower_blue02'    => 104,
          'royal_blue05'         => 105,
          'grass02'              => 106,
          'asparagus'            => 107,
          'swamp_green01'        => 108,
          'light_grey01'         => 109,
          'bluish03'             => 110,
          'cornflower_blue03'    => 111,
          'green_yellow01'       => 112,
          'pale_green02'         => 113,
          'emerald04'            => 114,
          'celadon'              => 115,
          'pale_blue01'          => 116,
          'sky02'                => 117,
          'viridian'             => 118,
          'vert_de_pomme04'      => 119,
          'mint_green02'         => 120,
          'chartreuse02'         => 121,
          'aquamarine04'         => 122,
          'electric02'           => 123,
          'bismarck_furious'     => 124,
          'eggplant02'           => 125,
          'red_violet01'         => 126,
          'violet_eggplant02'    => 127,
          'bright_violet'        => [128, 164],
          'violet04'             => 129,
          'ochre'                => 130,
          'pale_mauve02'         => 131,
          'pale_red_violet01'    => 132,
          'orchid01'             => 133,
          'amethyst03'           => 134,
          'heliotrope02'         => 135,
          'dark_goldenrod'       => 136,
          'pale_brown'           => 137,
          'mountbatten_pink'     => 138,
          'lilac01'              => 139,
          'wisteria01'           => 140,
          'amethyst04'           => 141,
          'old_gold'             => 142,
          'brass01'              => 143,
          'swamp_green02'        => 144,
          'light_grey02'         => 145,
          'niagara'              => 146,
          'wisteria02'           => 147,
          'lime01'               => 148,
          'pistachio01'          => 149,
          'moss_green'           => 150,
          'dark_tea_green'       => 151,
          'pale_blue02'          => 152,
          'pale_cornflower_blue' => 153,
          'green_yellow02'       => 154,
          'green_yellow03'       => 155,
          'pistachio02'          => 156,
          'chartreuse03'         => 157,
          'aquamarine05'         => 158,
          'pale_blue03'          => 159,
          'titian'               => 160,
          'cerise01'             => 161,
          'red_violet02'         => 162,
          'hot_pink01'           => 163,
          'magenta01'            => 165,
          'tenne'                => 166,
          'chestnut01'           => 167,
          'pale_red_violet02'    => 168,
          'orchid02'             => 169,
          'orchid03'             => 170,
          'heliotrope03'         => 171,
          'siena'                => 172,
          'dark_salmon'          => 173,
          'puce01'               => 174,
          'puce2'                => 175,
          'violaceous01'         => 176,
          'violaceous02'         => 177,
          'dark_pear'            => 178,
          'brass02'              => 179,
          'tan'                  => 180,
          'pale_chestnut'        => 181,
          'lilac02'              => 182,
          'wisteria03'           => 183,
          'childs_surprise'      => 184,
          'vert_de_peche'        => 185,
          'flax'                 => 186,
          'gray_tea_green'       => 187,
          'abdel_kerims_beard01' => 188,
          'periwinkle'           => 189,
          'lime02'               => 190,
          'pistachio03'          => 191,
          'pistachio04'          => 192,
          'tea_green01'          => 193,
          'tea_green02'          => 194,
          'pang'                 => 195,
          'crimson'              => 197,
          'cerise02'             => 198,
          'hot_pink02'           => 199,
          'hot_pink03'           => 200,
          'international_orange' => 202,
          'alizarin'             => 203,
          'dark_pink01'          => 204,
          'dark_pink02'          => 205,
          'shocked_star'         => 206,
          'fuchsia'              => 207,
          'tangerine'            => 208,
          'salmon'               => 209,
          'chestnut02'           => 210,
          'pale_red_violet03'    => 211,
          'pale_magenta'         => 212,
          'violaceous03'         => 213,
          'orange'               => 214,
          'pink_orange'          => 215,
          'saumon'               => 216,
          'pink01'               => 217,
          'pink02'               => 218,
          'violaceous04'         => 219,
          'gold'                 => 220,
          'mustard01'            => 221,
          'mustard02'            => 222,
          'dark_peach'           => 223,
          'pale_pink'            => 224,
          'thistle'              => 225,
          'corn01'               => 227,
          'corn02'               => 228,
          'perhydor'             => 229,
          'lemon_cream'          => 230,
          'black01'              => 232,
          'black02'              => 233,
          'black03'              => 234,
          'bistre'               => 235,
          'anthracite'           => 236,
          'wet_asphalt02'        => 237,
          'wet_asphalt03'        => 238,
          'wet_asphalt04'        => 239,
          'wet_asphalt05'        => 240,
          'wet_asphalt06'        => 241,
          'wet_asphalt07'        => 242,
          'gray02'               => 245,
          'gray03'               => 246,
          'gray04'               => 247,
          'gray05'               => 248,
          'light_grey03'         => 249,
          'light_grey04'         => 250,
          'silver01'             => 251,
          'abdel_kerims_beard02' => 252,
          'abdel_kerims_beard03' => 253,
          'abdel_kerims_beard04' => 254,
          'seashell'             => 255
        },
        16 => {
          'black'          => 30,
          'red'            => 31,
          'green'          => 32,
          'yellow'         => 33,
          'blue'           => 34,
          'magenta'        => 35,
          'cyan'           => 36,
          'white'          => 37,
          'bright_black'   => [30, 1],
          'bright_red'     => [31, 1],
          'bright_green'   => [32, 1],
          'bright_yellow'  => [33, 1],
          'bright_blue'    => [34, 1],
          'bright_magenta' => [35, 1],
          'bright_cyan'    => [36, 1],
          'bright_white'   => [37, 1],
        },
        8 => {
          'black'   => 30,
          'red'     => 31,
          'green'   => 32,
          'yellow'  => 33,
          'blue'    => 34,
          'magenta' => 35,
          'cyan'    => 36,
          'white'   => 37,
        }
      },
      :background => {
        16 => {
          'black'          => 40,
          'red'            => 41,
          'green'          => 42,
          'yellow'         => 43,
          'blue'           => 44,
          'magenta'        => 45,
          'cyan'           => 46,
          'white'          => 47,
          'bright_black'   => 40,
          'bright_red'     => 41,
          'bright_green'   => 42,
          'bright_yellow'  => 43,
          'bright_blue'    => 44,
          'bright_magenta' => 45,
          'bright_cyan'    => 46,
          'bright_white'   => 47,
        },
      }
    }
    @@colors[:background][256] = @@colors[:foreground][256]
    @@colors[:background][8]   = @@colors[:background][16]

    # The default colour options Hash.
    OPTS = {
      :from        => :readable,
      :foreground  => false,
      :background  => false,
      :bold        => false,
      :italic      => false,
      :underline   => false,
    }

    attr_reader :color_model

    # @return []
    attr_reader :options

    def initialize(color_model, options = {})
      @options     = OPTS.merge(options)
      @color_model = color_model
      set_layers
    end

    def foreground(readable = false)
      @readable_fg = find_color(:foreground)
      readable ? @readable_fg : layer_color(colors[@readable_fg])
    end

    def background(readable = false)
      @readable_bg = find_color(:background)
      readable ? @readable_bg : layer_color(colors(:background)[@readable_bg])
    end

    def to_ansi(inversion = false)
      ansi = create_ansi_sequence(foreground, background)
      escape(inversion ? ['7;', ansi].join('') : ansi)
    end

    private

    def set_layers
      foreground
      background
    end

    def escape(ansi)
      Array(ansi).map { |c| "\e[#{ c }m"}.join('')
    end

    def create_ansi_sequence(fg, bg, default_seq)
      (if fg && bg
        [[build_fg_sequence], [build_effects_sequence], [build_bg_sequence]].flatten
      elsif fg && !bg
        [[build_fg_sequence], [build_effects_sequence]].flatten
      elsif !fg && bg
        [[build_bg_sequence], [build_effects_sequence]].flatten
      else
        build_effects_sequence.tap { |sequence|
          sequence << default_seq if sequence.empty?
        }
      end).join(';')
    end

    def build_fg_sequence
      [foreground]
    end

    def build_effects_sequence
      [bold, italic, underline].delete_if { |e| e == false }
    end

    def build_bg_sequence
      [background]
    end

    def layer_color(layer)
      if layer
        layer.is_a?(Array) ? build_layer(layer) : layer
      else
        false
      end
    end

    def build_layer(layer)
      layer.first
    end

    def find_color(layer)
      case (color_id = cast_color(layer))
      when String
        return color_id if colors.has_key?(color_id)
      when Integer
        color_id = find_from_integer(color_id)
        return color_id if color_id
      when false
        return color_id
      end

      raise ArgumentError, %|invalid #{ layer.to_s } value "#{ color_id }"|
    end

    def cast_color(layer)
      return false unless options[layer]

      case options[:from]
      when :readable
        options[layer]
      when :hex
        HEX.new(options[layer]).to_term(color_model).to_i
      when :rgb
        RGB.new(options[layer]).to_term(color_model).to_i
      when :term
        TERM.new(options[layer], color_model).to_i
      end
    end

    def colors(layer = :foreground)
      @@colors[layer][color_model]
    end

    def sorted_colors
      colors.sort_by { |k, v| v.is_a?(Array) ? v.first + 10 : v }
    end

  end
end

Dir[File.expand_path('../colors/*.rb', __FILE__)].each do |file|
  require file
end
